tests/hw/esp32/variants.toit

```diff
@@ -0,0 +1,443 @@
+  i2c-pullup-test-pin -> int
+  i2c-pullup-other-pin -> int
+  i2c-pullup-measure-pin -> int
+
+  /*
+  2 Pins that are connected between the two boards.
+
+  The $connected-pin1 of board 1 is connected to the $connected-pin2 of board 2.
+  The $connected-pin2 of board 1 is connected to the $connected-pin1 of board 2.
+
+  Don't forget to connect the grounds as well.
+  */
+  connected-pin1 -> int
+  connected-pin2 -> int
+
+  /*
+  Pulse counter pins.
+
+  Connect $pulse-counter1-in1 to $pulse-counter1-out1 with a 330Ohm resistor.
+  Connect $pulse-counter1-in2 to $pulse-counter1-out2 with a 330Ohm resistor.
+  */
+  pulse-counter1-in1 -> int
+  pulse-counter1-out1 -> int
+
+  pulse-counter1-in2 -> int
+  pulse-counter1-out2 -> int
+
+  pulse-counter-channel-count -> int
+
+  /*
+  PWM pins.
+
+  Connect $pwm-in1 to $pwm-out1 with a 330Ohm resistor.
+  Connect $pwm-in2 to $pwm-out2 with a 330Ohm resistor.
+  */
+  pwm-in1 -> int
+  pwm-out1 -> int
+
+  pwm-in2 -> int
+  pwm-out2 -> int
+
+  /*
+  RMT pull-up test pins.
+
+  Same as $open-drain-test-pin and $open-drain-level-pin.
+  */
+  rmt-drain-pullup-test-pin -> int
+  rmt-drain-pullup-level-pin -> int
+  rmt-drain-pullup-measure-pin -> int
+
+  /*
+  RMT many test pins.
+
+  Connect $rmt-many-in1 to $rmt-many-test-out1 with a 330Ohm resistor.
+  Connect $rmt-many-in2 to $rmt-many-test-out2 with a 330Ohm resistor.
+  */
+  rmt-many-in1 -> int
+  rmt-many-out1 -> int
+
+  rmt-many-in2 -> int
+  rmt-many-out2 -> int
+
+  /*
+  RMT pins.
+
+  Connect $rmt-pin1 to $rmt-pin2 with a 330Ohm resistor.
+  Connect $rmt-pin2 to $rmt-pin3 with a 330Ohm resistor.
+  */
+  rmt-pin1 -> int
+  rmt-pin2 -> int
+  rmt-pin3 -> int
+
+  /*
+  SPI keep-active pins.
+
+  Connect $spi-keep-active-cs-pin to $spi-keep-active-in-cs-pin with a 330Ohm resistor.
+  */
+  spi-keep-active-cs-pin -> int
+  spi-keep-active-in-cs-pin -> int
+
+  /*
+  Touch pins.
+
+  Connect a jumper wire to the touch pins. These should otherwise be
+    unconnected.
+  $touch-pin1 should be yellow.
+  $touch-pin2 should be green.
+  */
+  touch-pin1 -> int
+  touch-pin2 -> int
+
+  /*
+  Uart baud-rate pins.
+
+  Connect uart-baud-rate-in1 to uart-baud-rate-out1 with a 330Ohm resistor.
+  Connect uart-baud-rate-in2 to uart-baud-rate-out2 with a 330Ohm resistor.
+  */
+  uart-baud-rate-in1 -> int
+  uart-baud-rate-out1 -> int
+
+  uart-baud-rate-in2 -> int
+  uart-baud-rate-out2 -> int
+
+  /*
+  Uart flush test pins.
+
+  Connect uart-flush-in1 to uart-flush-out1 with a 330Ohm resistor.
+  Connect uart-flush-in2 to uart-flush-out2 with a 330Ohm resistor.
+  */
+  uart-flush-in1 -> int
+  uart-flush-out1 -> int
+
+  uart-flush-in2 -> int
+  uart-flush-out2 -> int
+
+  /*
+  Wait-for-close test pins.
+
+  The $wait-for-close-pin should be connected to GND with an 1MOhm resistor.
+  */
+  wait-for-close-pin -> int
+
+VARIANTS ::= {
+  system.ARCHITECTURE-ESP32: Esp32,
+  system.ARCHITECTURE-ESP32S3: Esp32s3,
+}
+
+abstract class VariantBase:
```

Comment:
Does this implement `Variant`? Is it really necessary with both an abstract class and the separate interface?

Reply:
simplified.
done.